crates/pairwise_nostd_internal/src/accumulator.rs

```diff
@@ -30,86 +32,187 @@
+///   receive special treatment from the rust compiler to encourage
+///   optimizations with SIMD operations (I may be wrong). If this is indeed
+///   the case, I have some concerns that placing such types within a struct
+///   *could* interfere with this treatment (due to the struct's memory
+///   representation requirements).
+///
+/// With all that said, I think it's worthwhile use an explicit type for now.
+/// I think that there are significant advantages to improving the readability
+/// of the code and it would be relatively move away from using the struct
+/// representation later (it would be easier to do that before we start
+/// introducing more accumulators).
+pub struct DataElement {
+    pub value: f64,
+    pub weight: f64,
+}
 
-/// describes the output components from a single Accumulator statepack
+/// describes the output components from a single Accumulator accum_state
 pub enum OutputDescr {
     MultiScalarComp(&'static [&'static str]),
     SingleVecComp { size: usize, name: &'static str },
 }
 
 impl OutputDescr {
     /// the number of components to allocate per component
-    pub fn n_per_statepack(&self) -> usize {
+    pub fn n_per_accum_state(&self) -> usize {
         match self {
             Self::MultiScalarComp(names) => names.len(),
             Self::SingleVecComp { size, .. } => *size,
         }
     }
 }
 
-// Accumulators generally operate on individual state-packs (the implementation
-// is currently inefficient, but we will refactor and try to come up with
-// better abstractions once we are done mapping out all the requirements)
-//
-// In the context of the larger package, there will be n accumulators
+/// Accumulators generally operate on individual `accum_state`s.
+///
+/// The implementation is currently inefficient, but we will refactor and try
+/// to come up with better abstractions once we are done mapping out all the
+/// requirements.
+///
+/// # Note
+/// It might be elegant to:
+/// - introduce a separate, related trait, called `Consume` or `UpdateLeft`,
+///   which would look something like
+///   ```text
+///   pub trait UpdateLeft<Rhs> {
+///       fn update_left(accum_state: &mut ArrayViewMut1<f64>, right: Rhs);
+///   }
+///   ```
+/// - rename the current trait to something like `AccumMiscOps` and delete the
+///   `consume` and `merge` methods.
+/// - define the `Accumulator` trait as a super trait:
+///   ```text
+///   pub trait Accumulator:
+///       AccumMiscOps
+///       + UpdateLeft<&Self>
+///       + UpdateLeft<&DataElement>
+///   {}
+///   ```
+/// - thus for each Accumulator, we would be implementing `UpdateLeft<&Self>`
+///   (instead of `merge`) and `UpdateLeft<&DataElement>` (instead of
+///   `consume`).
+/// - The advantage to doing this is that `update_left` would be used for
+///   merges and consuming -- it's purely aesthetic. It may also be nice if we
+///   introduce the idea of partially digested data... (it probably isn't worth
+///   the effort)
```

Comment:
This is an interesting idea, but I think it's probably more abstraction than is warranted (as I think you are saying).

Reply:
yup, that's what I was saying. I'm going to leave the comment there, for now, on the off-chance that we decide it makes sense later